onnxruntime/core/providers/cuda/cuda_kernel.h

```diff
@@ -53,18 +54,43 @@ class CudaKernel : public OpKernel {
     return provider_->GetTransientScratchBuffer<T>(count_or_bytes);
   }
 
+  inline void AddDeferredReleaseCPUPtr(void* p, onnxruntime::Stream* ort_stream) const {
+    ORT_ENFORCE(ort_stream->device.Type() == OrtDevice::GPU);
+    auto* cuda_ep_stream = static_cast<CudaStream*>(ort_stream);
+    cuda_ep_stream->EnqueDeferredCPUBuffer(p);
+  }
+
   template <typename T>
   inline IAllocatorUniquePtr<T> AllocateBufferOnCPUPinned(size_t count_or_bytes) const {
     return provider_->AllocateBufferOnCPUPinned<T>(count_or_bytes);
   }
 
-  inline void AddDeferredReleaseCPUPtr(void* p) const {
-    provider_->AddDeferredReleaseCPUPtr(p);
+  const cudaDeviceProp& GetDeviceProp() const { return provider_->GetDeviceProp(); }
+
+  inline cudaStream_t Stream(OpKernelContext* ctx) const {
+    auto* stream = ctx->GetComputeStream();
+    return stream ? static_cast<cudaStream_t>(stream->handle) : nullptr;
   }
 
-  const cudaDeviceProp& GetDeviceProp() const { return provider_->GetDeviceProp(); }
+  inline cudnnHandle_t GetCudnnHandle(OpKernelContext* ctx) const {
+    return GetCudnnHandle(static_cast<CudaStream*>(ctx->GetComputeStream()));
+  }
 
-  inline cudaStream_t Stream() const { return static_cast<cudaStream_t>(provider_->GetComputeStream()); }
+  static inline cudnnHandle_t GetCudnnHandle(onnxruntime::CudaStream* stream) {
+    return stream->cudnn_handle_;
+  }
+
+  inline cublasHandle_t GetCublasHandle(OpKernelContext* ctx) const {
+    return GetCublasHandle(static_cast<CudaStream*>(ctx->GetComputeStream()));
+  }
+
+  static inline cublasHandle_t GetCublasHandle(onnxruntime::CudaStream* stream) {
+    return stream->cublas_handle_;
+  }
+
+  inline onnxruntime::Stream* OrtStream(OpKernelContext* ctx) const {
```

Comment:
Should return CudaStream, or move this function to the base class OpKernel

Reply:
my 2 cents, this function is not even necessary. If no polymorphic case, wherever OrtStream(context) is invoked, it can be replaced with context->GetComputeStream()